sdk/identity/azure-identity/src/managed_identity_source.cpp

```diff
@@ -250,20 +250,29 @@ std::unique_ptr<ManagedIdentitySource> AppServiceV2019ManagedIdentitySource::Cre
       credName, clientId, resourceId, options, "IDENTITY_ENDPOINT", "IDENTITY_HEADER", "2019");
 }
 
-// Cloud Shell doesn't support user-assigned managed identities
 std::unique_ptr<ManagedIdentitySource> CloudShellManagedIdentitySource::Create(
     std::string const& credName,
     std::string const& clientId,
-    std::string const&,
+    std::string const& resourceId,
     Azure::Core::Credentials::TokenCredentialOptions const& options)
 {
+  using Azure::Core::Credentials::AuthenticationException;
+
   constexpr auto EndpointVarName = "MSI_ENDPOINT";
   auto msiEndpoint = Environment::GetVariable(EndpointVarName);
 
   std::string const CredSource = "Cloud Shell";
 
   if (!msiEndpoint.empty())
   {
+    if (!clientId.empty() || !resourceId.empty())
+    {
+      throw AuthenticationException(
+          "User assigned identity is not supported in the Cloud Shell environment. To authenticate "
+          "with the system assigned identity, omit the client or resource ID when constructing the "
```

Comment:
I'm not sure it is technically correct that we are using the system assigned identity, but I'm not sure the distinction is important. @chlowell - I'm curious on your thoughts here.

Reply:
Seems like the User Identity (i.e. the signed-in user account) is typically authenticated (as evident by the UPN claim in the fetched token). Since system assigned identity generally implies a managed identity, I suppose this can be inaccurate?

Can probably just remove that part and say, "User-assigned managed identities are not supported in Cloud Shell environments. Omit the client or resource ID when constructing the ManagedIdentityCredential".

Reply:
The documentation in the .NET MICred ctor suggests not providing a client id means a system-assigned identity is selected:
https://github.com/Azure/azure-sdk-for-net/blob/05afb36edea3832eb86c3d169149bacd33b18436/sdk/identity/Azure.Identity/src/Credentials/ManagedIdentityCredential.cs#L36-L44

cc @schaabs

Reply:
System- vs. user-assigned isn't a useful distinction in describing a platform's default behavior. For example, a platform may have only a user-assigned identity and return tokens for that identity when a request doesn't specify an ID.

Also, Cloud Shell doesn't support managed identities of either kind. It always authenticates the signed-in user. Even so, I went with "the Cloud Shell API doesn't support user-assigned managed identities" as the error message for Go because that's true, probably reflects what the caller was trying to do, and I hope won't raise the question "if Cloud Shell doesn't support managed identities at all, why does `ManagedIdentityCredential` work in Cloud Shell?"